cpmpy/solvers/ortools.py

```diff
@@ -296,328 +210,321 @@ def solveAll(self, display=None, time_limit=None, solution_limit=None, **kwargs)
-        elif isinstance(cpm_var, _NumVarImpl): # _BoolVarImpl is subclass of _NumVarImpl
-            return self.varmap[cpm_var]
+        More specifically, it will branch that variable on that value first if possible. This is known as 'phase saving' in the SAT literature, but then extended to integer variables.
 
-        raise NotImplementedError("Not a know var {}".format(cpm_var))
+        The solution hint does NOT need to satisfy all constraints, it should just provide reasonable default values for the variables. It can decrease solving times substantially, especially when solving a similar model repeatedly
 
-    def ort_var_or_list(self, cpm_expr):
+        :param cpm_vars: list of CPMpy variables
+        :param vals: list of (corresponding) values for the variables
         """
-            like ort_var() but also works on lists of variables
+        self.ort_model.ClearHints()  # because add just appends
+        for (cpm_var, val) in zip(cpm_vars, vals):
+            self.ort_model.AddHint(self.solver_var(cpm_var), val)
+
 
-            Typically only needed for internal use
+    def get_core(self):
+        from ortools.sat.python import cp_model as ort
         """
-        if is_any_list(cpm_expr):
-            return [self.ort_var_or_list(sub) for sub in cpm_expr]
-        return self.ort_var(cpm_expr)
+            For use with s.solve(assumptions=[...]). Only meaningful if the solver returned UNSAT. In that case, get_core() returns a small subset of assumption variables that are unsat together.
 
+            CPMpy will return only those variables that are False (in the UNSAT core)
 
-    def ort_numexpr(self, cpm_expr):
-        """
-            converts CPMpy subexpression into corresponding
-            ORTools subexpressions (for in objective function and comparisons)
-            Accepted by ORTools:
-            - Decision variable: Var
-            - Linear: sum([Var])                                   (CPMpy class 'Operator', name 'sum')
-                      wsum([Const],[Var])                          (CPMpy class 'Operator', name 'wsum')
+            Note that there is no guarantee that the core is minimal, though this interface does upon up the possibility to add more advanced Minimal Unsatisfiabile Subset algorithms on top. All contributions welcome!
 
-            Typically only needed for internal use
+            For pure or-tools example, see http://github.com/google/or-tools/blob/master/ortools/sat/samples/assumptions_sample_sat.py
+
+            Requires or-tools >= 8.2!!!
         """
-        if is_num(cpm_expr):
-            return cpm_expr
+        assert (self.ort_status == ort.INFEASIBLE), "get_core(): solver must return UNSAT"
 
-        # decision variables, check in varmap
-        if isinstance(cpm_expr, _NumVarImpl): # _BoolVarImpl is subclass of _NumVarImpl
-            return self.ort_var(cpm_expr)
+        # use our own dict because of VarIndexToVarProto(0) bug in ort 8.2
```

Comment:
stray comment.

Your change broke it though: get_core() expects a list of CPMpy variables, but you return the ortools variables.

I don't think we have a way of going from the ort_var to the cpm_var. So perhaps the old 'own list' approach is still the best? (there never has been another implementation)

Reply:
Indeed, this should return CPMpy vars...
Also present in the solvers branch probably as I branched off that one...
Can't we retrieve the cpm_vars from the _varmap? They should be mapped there, although in the other direction...

Reply:
smth along the lines of:
```
ort_assum_vars = [self.ort_model.VarIndexToVarProto(i).name for i in assum_idx]
return [cpm_var for cpm_var in self._varmap if cpm_var.name in ort_assum_vars]
```